### PR TITLE
Added first draft of MS guest post on 1001 node cluster

### PR DESCRIPTION
**Pull request by ChrisChinchilla**

### CODE REVIEW

File: pages/a/ms-azure-guest-post.html [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+tags: a, b
+category: developernews
+
+# Crate guest post for Azure - Setting up Azure for Crate
+
+Crate.io has been striving to create the most scalable database in our industry. We aim to make scaling your data from millions to billions of records and beyond, as simple as possible.
+
+We have clients who write and query gigabytes of data a minute, store terabytes a day and love the performance and stability that Crate offers. But the question always remained at the back of our minds, when taken to the extreme, what is Crate really capable of?
+
+Our aims were to:
```
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: I'm not sure if these were the initial aims, but more aims we now have after the process. I'd love to know what our aims really were in the first place, in addition to just "wanting to know what Crate could do".

File: pages/a/ms-azure-guest-post.html [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+After a two day coding session with Microsoft staff in Vienna, they recommended we instead switched to using [Azure Templates](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authoring-templates/), a json formatted file that provisions all the resources and commands to be run for your application in a single operation. These templates can be used in conjunction with the Azure command line tool to rigger a complex cluster quickly, for example:
+
+```bash
+azure group deployment create -f azuredeploy.json \
+  -e azuredeploy.parameters.json \
+  -g <resourcegroup>
+```
+
+For the convenience of our team and to allow for reproduction of individual clusters, we broke these templates into cluster functions. I.e. one template for creating the network infrastructure, one for installing Docker components etc. (**Github link?**)
```
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: What's the final GitHub link and are we happy to publish it?

File: pages/a/ms-azure-guest-post.html [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+### Problems
+
+Plan A started well, but after a few hundred nodes we started to run into problems from different sources. First the network overhead of running traffic through the [Docker container bridge](https://docs.docker.com/engine/userguide/networking/dockernetworks/#a-bridge-network) proved to much and the Microsoft engineers recommended we switched to running Crate from Linux packages. Next there were problems with the Consul cluster as we had not configured it optimally to run on a multi-core machine and so it was struggling to meet the demand put upon it. After many hours debugging leading to a small config change (**What was it?**) we were able to get Consul running using all cores and reacting far more responsively. Finally despite being a great service, the Ruxit service couldn't cope with the traffic we were generating and the 29 metrics we wanted to collect. At around 500 nodes we were hitting their API so hard it shut down, so this wasn't a viable option for us, or for Ruxit.
```
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: Do we know what the change was and can we mention it?

File: pages/a/ms-azure-guest-post.html [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+With lessons learned it was time to make a few changes and start again, to summarize, this is what we changed:
+
+- Replaced Ruxit with [Ganglia](http://ganglia.info/) for monitoring the cluster.
+- Launch Crate from Linux packages instead of Docker containers
+
+### Problems
+
+The main issues encountered with this plan was the shear amount of connections and network traffic generated in a cluster of this size. For example, if the Consul, Docker Swarm or Crate master switches to a different node because of an issue, it will inform every node in the cluster, generating a lot of traffic across the network. The engineers at Microsoft were able to help us with this issue by changing some settings in the VM image kernel (**What was this?**) which improved cluster stability.
+
+This improved speeds and network stability and so we started to import our data set and hit upon a so far uncovered bug with our `COPY FROM` command (**What was it?**). Whilst the core team worked on solving this bug, we switched to using the [GitHub archive data set](https://www.githubarchive.org/) which offered more flexibility for import options so we could keep pushing forwards.
```
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: Can we mention what the problem was and if/when it was fixed?

File: pages/a/ms-azure-guest-post.html [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+## Next Steps
```
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: This section is a pre-pre-draft, more just notes, but feel free to add to it.

File: pages/a/ms-azure-guest-post.html [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+## Data Set
+
+70 terabytes of compressed data from the [common crawl database](http://commoncrawl.org/the-data/).
+
+## Plan A
+
+Our initial approach was to create a series of scripts to do setup the following:
+
+1. A 3 node [Consul](https://www.consul.io/) cluster to manage the cluster and handle service discovery.
+2. 3 Docker swarm master nodes to create VM instances quickly via docker machine.
```
Comment (kovrus) [inline pages/a/ms-azure-guest-post.html]: We have not been using docker machines at all. Instead, as alternative we have used **azure cli** for creating vms.

File: pages/a/ms-azure-guest-post.html [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+3. Gradually scale the cluster by creating docker instances running Crate and managed by the Swarm master.
+4. A 3 node monitoring and metrics cluster, we settled on [Ruxit](https://ruxit.com/).
```
Comment (kovrus) [inline pages/a/ms-azure-guest-post.html]: Ruxit agents were installed on each machine, they do not require a dedicated three nodes monitoring cluster

File: pages/a/ms-azure-guest-post.html [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+## Data Set
+
+70 terabytes of compressed data from the [common crawl database](http://commoncrawl.org/the-data/).
+
+## Plan A
+
+Our initial approach was to create a series of scripts to do setup the following:
+
+1. A 3 node [Consul](https://www.consul.io/) cluster to manage the cluster and handle service discovery.
```
Comment (kovrus) [inline pages/a/ms-azure-guest-post.html]: The very firsts attempt was with etcd instead of consul

File: pages/a/ms-azure-guest-post.html [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+The main issues encountered with this plan was the shear amount of connections and network traffic generated in a cluster of this size. For example, if the Consul, Docker Swarm or Crate master switches to a different node because of an issue, it will inform every node in the cluster, generating a lot of traffic across the network. The engineers at Microsoft were able to help us with this issue by changing some settings in the VM image kernel (**What was this?**) which improved cluster stability.
+
+This improved speeds and network stability and so we started to import our data set and hit upon a so far uncovered bug with our `COPY FROM` command. Whilst the core team worked on solving this bug, we switched to using the [GitHub archive data set](https://www.githubarchive.org/) which offered more flexibility for import options so we could keep pushing forwards.
+
+## Next Steps
+- RRD data extraction to analyze main issues between stability.
+- complete allocation of shards and partitions and cluster overhead.
+- How many shards on a node equals what in terms of performance.
+- Crate Master switched a lot, why?
```
Comment (kovrus) [inline pages/a/ms-azure-guest-post.html]: The election of a master node in the crate cluster seemed fine. it was quite a problem in a consul cluster.

File: pages/a/ms-azure-guest-post.html [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+- **Storage Account**: A data storage solution allocated to you that can contain Blob, Table, Queue, and File data. Can be stored on SSDs or HDDs.
+- **Virtual Network**: A representation of a network in the cloud, where you can define DHCP address blocks, DNS settings, security policies, and routing.
+- **Subnet**: Further segmentation for virtual networks.
+- **Virtual Machine (VM)**: Represents a particular operating system or application stack.
+- **Network Interface (NIC)**: Represents a network interface that can be associated to one or more virtual machines (VM).
+- **Network Security Group (NSG)**: Contains a list of rules that allow or deny network traffic to your VMs in a Virtual Network. Can be associated with subnets or individual VM instances within that subnet.
+
+## Data Set
+
+70 terabytes of compressed data from the [common crawl database](http://commoncrawl.org/the-data/).
```
Comment (kovrus) [inline pages/a/ms-azure-guest-post.html]: We have not imported much from common crowl. Is it worth mentioning?

File: pages/a/ms-azure-guest-post.html [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+Our aims were to:
+
+- See how Crate would distribute shards and partitions across such a large cluster
+- Examine the overhead added by a the cluster management
+- See what speeds the cluster could write data at
+- Another
+- And another
+
+Microsoft Azure were generously lent us the use of 8000 cores as the infrastructure for this test and engineer time to guide us through problems and questions we might have.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: Our extended limits were:
```
Name                        Unit   CurrentValue  Limit
--------------------------  -----  ------------  -----
Availability Sets           Count  0             2000
Total Regional Cores        Count  0             8000
Virtual Machines            Count  0             10000
Virtual Machine Scale Sets  Count  0             50
Standard Dv2 Family Cores   Count  0             8000
```

File: pages/a/ms-azure-guest-post.html [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+Microsoft hoped to prove that their Azure platform was reliable enough for a project of such rapid scaling and resource intensive.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: "hoped to prove" sounds a bit weak. they wanted to show the capabilities and reliability of their platform. maybe jodok can elaborate a bit more on that.

File: pages/a/ms-azure-guest-post.html [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+So, what happened and how did we do it?
+
+We aimed to create a 1001 node Crate cluster and xxx. Setting up such a cluster on any kind of infrastructure requires planning and whilst cloud hosts simplify this process, understanding how to translate your requirements into their paradigms requires research and experimentation.
+
+Azure has several concepts that we needed to enable and configure to get our cluster started:
+
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: * Resource [P]
  * Deployment - a set of (referencing and depending) resources that should be deployed on the platform
* Network [P]
  * VirtualNetwork [R]
    * Subnet (property of VirtualNetwork)
  * NetworkInterface [R]
     IPConfiguration (property of NetworkInterface)
  * PublicIPAddress [R]
  * NetworkSecurityGroup [R]
    * Rules (inbound and outbound, property of NSG)
* Storage [P]
  * StorageAccount [R]
* Compute [P]
  * AvailabilitySet \[R\] (make sure vms are scheduled on correct hosts, [read](https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-manage-availability/#avoid-single-instance-virtual-machines-in-availability-sets))
  * VirtualMachine [R]
  * VirtualMachineExtension [R]

[P] are Providers (you can list them using the cli `azure provider list`)
[R] are Resources

File: pages/a/ms-azure-guest-post.html (hunk 8 above)
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: ... however, ended up with consul.

File: pages/a/ms-azure-guest-post.html [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+Our initial approach was to create a series of scripts to do setup the following:
+
+1. A 3 node [Consul](https://www.consul.io/) cluster to manage the cluster and handle service discovery.
+2. 3 Docker swarm master nodes to create VM instances quickly via docker machine.
+3. Gradually scale the cluster by creating docker instances running Crate and managed by the Swarm master.
+4. A 3 node monitoring and metrics cluster, we settled on [Ruxit](https://ruxit.com/).
+5. Import our data set using Crate's `Copy From` command.
+
+After a two day coding session with Microsoft staff in Vienna, they recommended we instead switched to using [Azure Templates](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authoring-templates/), a json formatted file that provisions all the resources and commands to be run for your application in a single operation. These templates can be used in conjunction with the Azure command line tool to rigger a complex cluster quickly, for example:
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: rigger -> trigger

File: pages/a/ms-azure-guest-post.html [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+```bash
+azure group deployment create -f azuredeploy.json \
+  -e azuredeploy.parameters.json \
+  -g <resourcegroup>
+```
+
+For the convenience of our team and to allow for reproduction of individual clusters, we broke these templates into cluster functions. I.e. one template for creating the network infrastructure, one for installing Docker components etc.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: cluster functions -> logical self contained units

File: pages/a/ms-azure-guest-post.html [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+### Problems
+
+Plan A started well, but after a few hundred nodes we started to run into problems from different sources. First the network overhead of running traffic through the [Docker container bridge](https://docs.docker.com/engine/userguide/networking/dockernetworks/#a-bridge-network) proved to much and the Microsoft engineers recommended we switched to running Crate from Linux packages. Next there were problems with the Consul cluster as we had not configured it optimally to run on a multi-core machine and so it was struggling to meet the demand put upon it. After many hours debugging leading to a small config change we were able to get Consul running using all cores and reacting far more responsively. Finally despite being a great service, the Ruxit service couldn't cope with the traffic we were generating and the 29 metrics we wanted to collect. At around 500 nodes we were hitting their API so hard it shut down, so this wasn't a viable option for us, or for Ruxit.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: the actual problem was, that using docker swarm we had 3 distributed systems running:
* a cluster for consul - persitance backend for swarm
* docker swarm
* crate cluster

running the consul cluster at this scale already resulted in a lot of open connections. on top of it, each swarm node reported to consul.
starting crate then, relies on a stable swarm/consul cluster (which was not really the case), but once crate was starting up, additional connections between all nodes are opened.

after making consul using multiple cpu cores (setting `GOMAXPROCS` environment variable to num cores) it became more stable.

running crate from a naked installation was our own intention, to prove that it is actually not crate causing the network problems.

Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: we should also mention, why Ruxit cannot keep up this large scale:
the reason is that they monitor each process and the connections between processes, which leads in our case to a massive amount of data. @joemoe knows more about it

Comment (joemoe) [inline pages/a/ms-azure-guest-post.html]: @ChrisChinchilla ruxit and docker shouldn't come out too bad in this post.

File: pages/a/ms-azure-guest-post.html [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+This improved speeds and network stability and so we started to import our data set and hit upon a so far uncovered bug with our `COPY FROM` command. Whilst the core team worked on solving this bug, we switched to using the [GitHub archive data set](https://www.githubarchive.org/) which offered more flexibility for import options so we could keep pushing forwards.
+
+## Next Steps
+- RRD data extraction to analyze main issues between stability.
+- complete allocation of shards and partitions and cluster overhead.
+- How many shards on a node equals what in terms of performance.
+- Crate Master switched a lot, why?
+
+We can't say this was production ready and made lots of queries, no, not possible. Wasn't stable enough.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: that sounds like a very bad outcome :(
although there were quite a lot of obstacles, we have gained a lot of knowledge about running crate at very large scale.

**Comment by chaudum** [comment]
one sentence per line would be good for commenting :)

File: pages/a/ms-azure-guest-post.html (hunk 16 above)
Comment (ChrisChinchilla) [inline pages/a/ms-azure-guest-post.html]: @chaudum Was the Docker Bridge issue mentioned ever a problem?

File: pages/a/ms-azure-guest-post.html [hunk 18]
```diff
@@ -13,47 +13,57 @@
...
-- And another
+- Know how Crate would distribute shards and partitions across such a large cluster.
+- Examine the overhead added by a the cluster management.
+- See at what speeds the cluster could write data.
 
-Microsoft Azure were generously lent us the use of 8000 cores as the infrastructure for this test and engineer time to guide us through problems and questions we might have.
+Microsoft Azure generously lent us the use of 8000 cores as the infrastructure for this test and engineer time to guide us through problems and questions we might have.
 
-Microsoft hoped to prove that their Azure platform was reliable enough for a project of such rapid scaling and resource intensive.
+Microsoft wanted to show that their Azure platform was reliable and capable enough for a project that would scale so rapidly and be so resource intesive.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: typo: intensive

Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: that would scale so rapidly => at such a scale

File: pages/a/ms-azure-guest-post.html [hunk 19]
```diff
@@ -13,47 +13,57 @@
...
 
+And some limitations that influenced our architecture decisions:
+
+- 8000 CPU cores
+- 10000 VMs
+- 100 VMs in single availability set
+- 800 resources in single resource group
+- 4096 IP addresses in a vnet
+
+This led us to create what we termed a 'scaling unit' that contained all the VMS and resources required to support 100 Crate nodes. This meant we had 10 scaling units across the cluster.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: the term "scale unit" comes from MS

File: pages/a/ms-azure-guest-post.html [hunk 20]
```diff
@@ -13,47 +13,57 @@
...
-Azure has several concepts that we needed to enable and configure to get our cluster started:
+Azure has several concepts that we needed to understand and make use of to get our cluster started:
 
 - **Resource Group**: A 'container' that holds related resources of all types for an application.
 - **Storage Account**: A data storage solution allocated to you that can contain Blob, Table, Queue, and File data. Can be stored on SSDs or HDDs.
-- **Virtual Network**: A representation of a network in the cloud, where you can define DHCP address blocks, DNS settings, security policies, and routing.
+- **Virtual Network (vnet)**: A representation of a network in the cloud, where you can define DHCP address blocks, DNS settings, security policies, and routing.
 - **Subnet**: Further segmentation for virtual networks.
 - **Virtual Machine (VM)**: Represents a particular operating system or application stack.
 - **Network Interface (NIC)**: Represents a network interface that can be associated to one or more virtual machines (VM).
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: a single NIC can only be assigned to a single VM
but a VM can have multiple NICs assigned

File: pages/a/ms-azure-guest-post.html [hunk 21]
```diff
@@ -13,47 +13,57 @@
...
 
-Azure has several concepts that we needed to enable and configure to get our cluster started:
+Azure has several concepts that we needed to understand and make use of to get our cluster started:
 
 - **Resource Group**: A 'container' that holds related resources of all types for an application.
 - **Storage Account**: A data storage solution allocated to you that can contain Blob, Table, Queue, and File data. Can be stored on SSDs or HDDs.
-- **Virtual Network**: A representation of a network in the cloud, where you can define DHCP address blocks, DNS settings, security policies, and routing.
+- **Virtual Network (vnet)**: A representation of a network in the cloud, where you can define DHCP address blocks, DNS settings, security policies, and routing.
 - **Subnet**: Further segmentation for virtual networks.
 - **Virtual Machine (VM)**: Represents a particular operating system or application stack.
```
Comment (chaudum) [inline pages/a/ms-azure-guest-post.html]: VM = emulation of a complete computer system to a level that is required to fully run an operating system

**Pull request closed without merging**